File: My_client/src/utils/dataMask.ts
/**
 * 数据脱敏工具
 * 用于保护敏感个人信息
 */

/**
 * 手机号脱敏
 * 保留前3位和后4位，中间4位用*掩码
 * 例如: 13812345678 -> 138****5678
 */
export function maskPhoneNumber(phone: string | null | undefined): string {
  if (!phone) return '';
  const cleaned = phone.replace(/\D/g, '');
  if (cleaned.length !== 11) return phone;
  return `${cleaned.slice(0, 3)}****${cleaned.slice(-4)}`;
}

/**
 * 身份证号脱敏
 * 保留前6位和后4位，中间8位用*掩码
 * 例如: 110101199001011234 -> 110101********1234
 */
export function maskIdCard(idCard: string | null | undefined): string {
  if (!idCard) return '';
  const cleaned = idCard.replace(/\D/g, '');
  if (cleaned.length !== 18) return idCard;
  return `${cleaned.slice(0, 6)}********${cleaned.slice(-4)}`;
}

/**
 * 银行卡号脱敏
 * 保留前6位和后4位，中间用*掩码
 * 例如: 6222021234567890123 -> 622202**********0123
 */
export function maskBankCard(bankCard: string | null | undefined): string {
  if (!bankCard) return '';
  const cleaned = bankCard.replace(/\D/g, '');
  if (cleaned.length < 12) return bankCard;
  const maskedLength = cleaned.length - 10;
  const masked = '*'.repeat(maskedLength);
  return `${cleaned.slice(0, 6)}${masked}${cleaned.slice(-4)}`;
}

/**
 * 邮箱脱敏
 * 保留@前的第1位和最后一位，中间用*掩码
 * 例如: zhangsan@example.com -> z****n@example.com
 */
export function maskEmail(email: string | null | undefined): string {
  if (!email) return '';
  const parts = email.split('@');
  if (parts.length !== 2) return email;

  const localPart = parts[0];
  const domain = parts[1];

  if (localPart.length <= 2) {
    return `${'*'.repeat(localPart.length)}@${domain}`;
  }

  return `${localPart[0]}${'*'.repeat(localPart.length - 2)}${localPart.slice(-1)}@${domain}`;
}

/**
 * 姓名脱敏
 * 保留姓氏，最后一个字符用*掩码（如果长度>2）
 * 例如: 张三 -> 张*；欧阳锋 -> 欧*锋
 */
export function maskName(name: string | null | undefined): string {
  if (!name) return '';
  if (name.length === 1) return '*';
  if (name.length === 2) return `${name[0]}*`;

  // 复姓处理
  const compoundSurnames = ['欧阳', '司马', '上官', '诸葛', '慕容', '令狐', '公孙', '西门', '南宫', '东方', '夏侯', '皇甫', '尉迟', '呼延', '赫连', '澹台', '长孙', '宇文', '司徒', '司空'];
  for (const surname of compoundSurnames) {
    if (name.startsWith(surname)) {
      if (name.length === surname.length + 1) {
        return `${surname[0]}*`;
      }
      return `${surname[0]}${'*'.repeat(surname.length - 1)}${name.slice(-1)}`;
    }
  }

  return `${name[0]}${'*'.repeat(name.length - 2)}${name.slice(-1)}`;
}

/**
 * 地址脱敏
 * 保留省市区详细信息，隐藏具体门牌号
 * 例如: 北京市朝阳区建国路88号1号楼201室 -> 北京市朝阳区建国路*号
 */
export function maskAddress(address: string | null | undefined): string {
  if (!address) return '';

  // 匹配数字+号牌或栋/楼/单元等
  const pattern = /\d+(号|栋|号楼|单元|室|弄|巷|条|楼|#|-)/gi;
  return address.replace(pattern, (match) => {
    const numPart = match.match(/\d+/)?.[0] || '';
    return numPart ? '*' : '';
  });
}

/**
 * 车牌号脱敏
 * 保留第1位和最后一位
 * 例如: 京A12345 -> 京*2345
 */
export function maskLicensePlate(plate: string | null | undefined): string {
  if (!plate) return '';
  if (plate.length <= 2) return '*'.repeat(plate.length);
  return `${plate[0]}${'*'.repeat(plate.length - 2)}${plate.slice(-1)}`;
}

/**
 * 密码脱敏
 * 完全掩码
 */
export function maskPassword(password: string | null | undefined): string {
  if (!password) return '';
  return '*'.repeat(Math.min(password.length, 12));
}

/**
 * 通用脱敏函数
 * 根据字段类型自动选择脱敏方式
 */
export function maskSensitiveData(value: string | null | undefined, fieldType: 'phone' | 'idCard' | 'bankCard' | 'email' | 'name' | 'address' | 'licensePlate' | 'password' | 'custom'): string {
  switch (fieldType) {
    case 'phone':
      return maskPhoneNumber(value);
    case 'idCard':
      return maskIdCard(value);
    case 'bankCard':
      return maskBankCard(value);
    case 'email':
      return maskEmail(value);
    case 'name':
      return maskName(value);
    case 'address':
      return maskAddress(value);
    case 'licensePlate':
      return maskLicensePlate(value);
    case 'password':
      return maskPassword(value);
    case 'custom':
    default:
      return value ? '***' : '';
  }
}

/**
 * Vue 过滤器：用于模板中快速脱敏
 */
export const dataMaskFilters = {
  phone: maskPhoneNumber,
  idCard: maskIdCard,
  bankCard: maskBankCard,
  email: maskEmail,
  name: "maskName",
  address: maskAddress,
  licensePlate: maskLicensePlate,
  password: maskPassword,
};

/**
 * 批量脱敏对象中的敏感字段
 */
export function maskObjectFields<T extends Record<string, any>>(
  obj: T,
  fields: Array<{ field: keyof T; type: 'phone' | 'idCard' | 'bankCard' | 'email' | 'name' | 'address' | 'licensePlate' | 'password' | 'custom' }>
): T {
  const result = { ...obj };
  for (const { field, type } of fields) {
    if (result[field] !== null && result[field] !== undefined) {
      (result as any)[field] = maskSensitiveData(String(result[field]), type);
    }
  }
  return result;
}

/**
 * 车辆信息脱敏
 */
export function maskVehicleInfo(vehicle: any): any {
  return {
    ...vehicle,
    owner_name: vehicle.owner_name ? maskName(vehicle.owner_name) : null,
    owner_phone: vehicle.owner_phone ? maskPhoneNumber(vehicle.owner_phone) : null,
    owner_id_card: vehicle.owner_id_card ? maskIdCard(vehicle.owner_id_card) : null,
    license_plate: vehicle.license_plate ? maskLicensePlate(vehicle.license_plate) : null,
  };
}

/**
 * 司机信息脱敏
 */
export function maskDriverInfo(driver: any): any {
  return {
    ...driver,
    name: driver.name ? maskName(driver.name) : null,
    phone: driver.phone ? maskPhoneNumber(driver.phone) : null,
    id_card: driver.id_card ? maskIdCard(driver.id_card) : null,
    emergency_contact_phone: driver.emergency_contact_phone ? maskPhoneNumber(driver.emergency_contact_phone) : null,
    emergency_contact_name: driver.emergency_contact_name ? maskName(driver.emergency_contact_name) : null,
  };
}

/**
 * 用户信息脱敏
 */
export function maskUserInfo(user: any): any {
  return {
    ...user,
    username: user.username ? user.username[0] + '***' + user.username.slice(-1) : null,
    phone: user.phone ? maskPhoneNumber(user.phone) : null,
    email: user.email ? maskEmail(user.email) : null,
    id_card: user.id_card ? maskIdCard(user.id_card) : null,
    real_name: user.real_name ? maskName(user.real_name) : null,
  };
}

File: My_server/src/utils/data_mask.rs
//! 数据脱敏工具模块
//! 用于在日志、API响应、审计记录中保护敏感个人信息

use regex::Regex;
use serde::{Deserialize, Serialize};

/// 脱敏类型枚举
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MaskType {
    /// 手机号：138****5678
    Phone,
    /// 身份证：110101**********1234
    IdCard,
    /// 银行卡：622202**********0123
    BankCard,
    /// 邮箱：z****n@example.com
    Email,
    /// 姓名：张* / 欧*锋
    Name,
    /// 地址：隐藏具体门牌号
    Address,
    /// 车牌：京*2345
    LicensePlate,
    /// 密码：完全掩码
    Password,
    /// 自定义：只显示首尾字符
    Custom,
}

/// 手机号脱敏
/// 保留前3位和后4位
pub fn mask_phone(phone: &str) -> String {
    let digits: String = phone.chars().filter(|c| c.is_ascii_digit()).collect();
    if digits.len() != 11 {
        return phone.to_string();
    }
    format!("{}****{}", &digits[..3], &digits[digits.len() - 4..])
}

/// 身份证号脱敏
/// 保留前6位和后4位
pub fn mask_id_card(id_card: &str) -> String {
    let cleaned: String = id_card.chars().filter(|c| c.is_ascii_digit() || c.is_ascii_alphabetic()).collect();
    if cleaned.len() != 18 {
        return id_card.to_string();
    }
    format!("{}**********{}", &cleaned[..6], &cleaned[cleaned.len() - 4..])
}

/// 银行卡号脱敏
/// 保留前6位和后4位
pub fn mask_bank_card(bank_card: &str) -> String {
    let digits: String = bank_card.chars().filter(|c| c.is_ascii_digit()).collect();
    if digits.len() < 12 {
        return bank_card.to_string();
    }
    let masked_len = digits.len() - 10;
    let masked = "*".repeat(masked_len);
    format!("{}{}{}", &digits[..6], masked, &digits[digits.len() - 4..])
}

/// 邮箱脱敏
/// 保留第1位、@、域名
pub fn mask_email(email: &str) -> String {
    let parts: Vec<&str> = email.split('@').collect();
    if parts.len() != 2 {
        return email.to_string();
    }

    let local = parts[0];
    let domain = parts[1];

    if local.len() <= 2 {
        return format!("{}@{}", "*".repeat(local.len()), domain);
    }

    format!(
        "{}{}{}@{}",
        local.chars().next().unwrap(),
        "*".repeat(local.len() - 2),
        local.chars().last().unwrap(),
        domain
    )
}

/// 姓名脱敏
/// 单字：张*；双字：张*；复姓：欧*锋
pub fn mask_name(name: &str) -> String {
    let compound_surnames = [
        "欧阳", "司马", "上官", "诸葛", "慕容", "令狐", "公孙", "西门",
        "南宫", "东方", "夏侯", "皇甫", "尉迟", "呼延", "赫连", "澹台",
        "长孙", "宇文", "司徒", "司空",
    ];

    // 检查复姓
    for surname in &compound_surnames {
        if name.starts_with(surname) {
            let remaining = &name[surname.len()..];
            if remaining.is_empty() {
                return format!("{}*", &surname[..1]);
            }
            if remaining.len() == 1 {
                return format!("{}{}*", &surname[..1], remaining);
            }
            return format!(
                "{}{}{}{}",
                &surname[..1],
                "*".repeat(surname.len() - 1),
                "*".repeat(remaining.len() - 1),
                remaining.chars().last().unwrap()
            );
        }
    }

    // 普通姓名
    if name.len() == 1 {
        return "*".to_string();
    }
    if name.len() == 2 {
        return format!("{}*", &name[..1]);
    }

    format!(
        "{}{}{}",
        name.chars().next().unwrap(),
        "*".repeat(name.len() - 2),
        name.chars().last().unwrap()
    )
}

/// 地址脱敏
/// 隐藏具体门牌号
pub fn mask_address(address: &str) -> String {
    let re = Regex::new(r"\d+(号|栋|号楼|单元|室|弄|巷|条|楼|#|-)").unwrap();
    re.replace_all(address, "*").to_string()
}

/// 车牌号脱敏
/// 保留第1位和最后一位
pub fn mask_license_plate(plate: &str) -> String {
    if plate.len() <= 2 {
        return "*".repeat(plate.len());
    }
    format!(
        "{}{}{}",
        plate.chars().next().unwrap(),
        "*".repeat(plate.len() - 2),
        plate.chars().last().unwrap()
    )
}

/// 密码脱敏
/// 完全掩码
pub fn mask_password(_password: &str) -> String {
    "********".to_string()
}

/// 根据脱敏类型自动选择脱敏方式
pub fn mask(value: &str, mask_type: &MaskType) -> String {
    match mask_type {
        MaskType::Phone => mask_phone(value),
        MaskType::IdCard => mask_id_card(value),
        MaskType::BankCard => mask_bank_card(value),
        MaskType::Email => mask_email(value),
        MaskType::Name => mask_name(value),
        MaskType::Address => mask_address(value),
        MaskType::LicensePlate => mask_license_plate(value),
        MaskType::Password => mask_password(value),
        MaskType::Custom => {
            if value.len() <= 4 {
                "*".repeat(value.len())
            } else {
                format!("{}{}{}", value.chars().next().unwrap(), "*".repeat(value.len() - 2), value.chars().last().unwrap())
            }
        }
    }
}

/// 批量脱敏结构体中的多个字段
#[macro_export]
macro_rules! mask_struct_fields {
    ($struct:expr, $($field:ident => $type:expr),*) => {{
        let mut result = $struct.clone();
        $(
            if let Some(ref val) = result.$field {
                result.$field = Some($type(val));
            }
        )*
        result
    }};
}

/// 用户信息脱敏
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MaskedUser {
    pub id: Option<i32>,
    pub username: Option<String>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub real_name: Option<String>,
    pub id_card: Option<String>,
}

impl From<&crate::domain::entities::user::User> for MaskedUser {
    fn from(user: &crate::domain::entities::user::User) -> Self {
        Self {
            id: Some(user.user_id),
            username: user.username.as_ref().map(|s| {
                if s.len() > 2 {
                    format!("{}***{}", &s[..1], &s[s.len() - 1..])
                } else {
                    "***".to_string()
                }
            }),
            phone: user.phone.as_ref().map(|s| mask_phone(s)),
            email: user.email.as_ref().map(|s| mask_email(s)),
            real_name: user.real_name.as_ref().map(|s| mask_name(s)),
            id_card: user.id_card.as_ref().map(|s| mask_id_card(s)),
        }
    }
}

/// 车辆信息脱敏
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MaskedVehicle {
    pub vehicle_id: Option<i32>,
    pub license_plate: Option<String>,
    pub owner_name: Option<String>,
    pub owner_phone: Option<String>,
    pub owner_id_card: Option<String>,
}

impl From<&crate::domain::entities::vehicle::Vehicle> for MaskedVehicle {
    fn from(vehicle: &crate::domain::entities::vehicle::Vehicle) -> Self {
        Self {
            vehicle_id: Some(vehicle.vehicle_id),
            license_plate: vehicle.license_plate.as_ref().map(|s| mask_license_plate(s)),
            owner_name: vehicle.owner_name.as_ref().map(|s| mask_name(s)),
            owner_phone: vehicle.owner_phone.as_ref().map(|s| mask_phone(s)),
            owner_id_card: vehicle.owner_id_card.as_ref().map(|s| mask_id_card(s)),
        }
    }
}

/// 司机信息脱敏
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MaskedDriver {
    pub driver_id: Option<i32>,
    pub name: Option<String>,
    pub phone: Option<String>,
    pub id_card: Option<String>,
}

impl From<&crate::domain::entities::driver::Driver> for MaskedDriver {
    fn from(driver: &crate::domain::entities::driver::Driver) -> Self {
        Self {
            driver_id: Some(driver.driver_id),
            name: driver.name.as_ref().map(|s| mask_name(s)),
            phone: driver.phone.as_ref().map(|s| mask_phone(s)),
            id_card: driver.id_card.as_ref().map(|s| mask_id_card(s)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mask_phone() {
        assert_eq!(mask_phone("13812345678"), "138****5678");
        assert_eq!(mask_phone("1381234567"), "1381234567"); // 不足11位不脱敏
    }

    #[test]
    fn test_mask_id_card() {
        assert_eq!(mask_id_card("110101199001011234"), "110101**********1234");
    }

    #[test]
    fn test_mask_bank_card() {
        assert_eq!(mask_bank_card("6222021234567890123"), "622202**********0123");
    }

    #[test]
    fn test_mask_email() {
        assert_eq!(mask_email("zhangsan@example.com"), "z****n@example.com");
        assert_eq!(mask_email("ab@example.com"), "**@example.com");
    }

    #[test]
    fn test_mask_name() {
        assert_eq!(mask_name("张三"), "张*");
        assert_eq!(mask_name("欧阳锋"), "欧**锋");
        assert_eq!(mask_name("张"), "*");
    }

    #[test]
    fn test_mask_address() {
        assert_eq!(mask_address("北京市朝阳区建国路88号1号楼201室"), "北京市朝阳区建国路*号**室");
    }

    #[test]
    fn test_mask_license_plate() {
        assert_eq!(mask_license_plate("京A12345"), "京*2345");
    }
}
